tapdb/sqlc/migrations/000020_asset_unique_key.up.sql

```diff
@@ -1,4 +1,99 @@
+-- Step 1: If the assets were spent, some of the duplicates might not have been
+-- updated on that flag. To make sure we can properly group on the spent flag
+-- below, we now update all assets that are spent.
+UPDATE assets
+SET spent = true
+WHERE asset_id IN (SELECT a.asset_id
+                   FROM assets a
+                            JOIN managed_utxos mu
+                                 ON a.anchor_utxo_id = mu.utxo_id
+                            JOIN chain_txns ct
+                                 ON mu.txn_id = ct.txn_id
+                            LEFT JOIN asset_transfer_inputs ati
+                                      ON ati.anchor_point = mu.outpoint
+                   WHERE a.spent = false
+                     AND ati.input_id IS NOT NULL);
+
+-- Step 2: Create a temporary table to store the minimum asset_id for each
+-- unique combination.
+CREATE TABLE tmp_min_assets AS
+SELECT MIN(asset_id) AS min_asset_id,
+       genesis_id,
+       script_key_id,
+       amount,
+       anchor_utxo_id,
+       spent
+FROM assets
+GROUP BY genesis_id, script_key_id, amount, anchor_utxo_id, spent;
+
+-- Step 3: Create a mapping table to track old and new asset_ids.
+CREATE TABLE tmp_asset_id_mapping AS
+SELECT a.asset_id       AS old_asset_id,
+       tmp.min_asset_id AS new_asset_id
+FROM assets a
+         JOIN tmp_min_assets tmp
+              ON a.genesis_id = tmp.genesis_id
+                  AND a.script_key_id = tmp.script_key_id
+                  AND a.amount = tmp.amount
+                  AND a.anchor_utxo_id = tmp.anchor_utxo_id
+                  AND a.spent = tmp.spent;
+
+-- Step 4: To make the next step possible, we need to disable a unique index on
+-- the asset_witnesses table. We'll re-create it later.
+DROP INDEX IF EXISTS asset_witnesses_asset_id_witness_index_unique;
+
+-- Step 5: Update the asset_witnesses and asset_proofs tables to reference the
+-- new asset_ids.
+UPDATE asset_witnesses
+SET asset_id = tmp_asset_id_mapping.new_asset_id
+FROM tmp_asset_id_mapping
+WHERE asset_witnesses.asset_id = tmp_asset_id_mapping.old_asset_id;
+
+-- For the proofs we need skip re-assigning them to the asset that we're going
+-- to keep if it already has a proof. This is because the unique index on the
+-- asset_proofs table would prevent us from doing so. And we can't disable the
+-- unique index, because it is an unnamed/inline index.
+UPDATE asset_proofs
+SET asset_id = filtered_mapping.new_asset_id
+FROM (                  
+        SELECT MIN(old_asset_id) AS old_asset_id, new_asset_id
```

Comment:
`asset_id` in table `asset_proofs` has a `UNIQUE` constraint. So if we map `old_asset_id` a and b both to `new_asset_id` x we violate that constraint. So by grouping on `new_asset_id` and taking `MIN(old_asset_id)` we only map a single `old_asset_id` to the `new_asset_id`. (a to x in this case) The remaining asset_proof with foreign key b will be deleted in step 8.

The `filtered_mapping` table does the exact same thing as `tmp_asset_id_mapping` but taking into account the `UNIQUE` constraint on `asset_id` in table `asset_proofs`. Hence I reused the column names.

> So the end result would be that, for duplicate assets, if the proof was not already using the new_asset_id (which is the smaller asset ID), asset_proofs.asset_id would be updated to use that new_asset_id.

Correct.

> And any proof already referencing that new_asset_id should be unchanged.

It will get updated, but with the same values, so the result is unchanged.